File: rust/tests/rfid/mfrc522_test/src/main.rs
use embedded_hal_bus::spi::ExclusiveDevice;
use linux_embedded_hal::SpidevBus;
use periph::chips::rfid::{Mfrc522Full, RX_GAIN_18_DB, RX_GAIN_23_DB, RX_GAIN_33_DB, RX_GAIN_38_DB, RX_GAIN_43_DB, RX_GAIN_48_DB};
use spidev::{SpiModeFlags, Spidev, SpidevOptions};

struct NullCs;
impl embedded_hal::digital::ErrorType for NullCs {
    type Error = core::convert::Infallible;
}
impl embedded_hal::digital::OutputPin for NullCs {
    fn set_low(&mut self) -> Result<(), Self::Error> { Ok(()) }
    fn set_high(&mut self) -> Result<(), Self::Error> { Ok(()) }
}

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let dev_path = std::env::var("SPI_DEV").unwrap_or_else(|_| "/dev/spidev0.0".to_string());

    let mut passed = 0i32;
    let mut failed = 0i32;

    let mut spi = Spidev::open(dev_path).expect("open spi");
    spi.configure(&SpidevOptions::new()
        .max_speed_hz(1_000_000)
        .mode(SpiModeFlags::SPI_MODE_0)
        .build()).expect("configure spi");
    let bus = SpidevBus(spi);
    let device = ExclusiveDevice::new_no_delay(bus, NullCs).expect("spi device");

    let mut mfrc = Mfrc522Full::new(device).expect("init MFRC522");

    let (chip, ver) = mfrc.version().unwrap_or((0, 0));
    check_true!(chip == 0x09, "chip_type == 0x09 (MFRC522)", passed, failed);
    check_true!(ver == 1 || ver == 2, "version in {1, 2}", passed, failed);

    mfrc.antenna_on().expect("antenna_on");
    mfrc.antenna_off().expect("antenna_off");
    mfrc.antenna_on().expect("antenna_on");

    for gain in [RX_GAIN_18_DB, RX_GAIN_23_DB, RX_GAIN_33_DB, RX_GAIN_38_DB, RX_GAIN_43_DB, RX_GAIN_48_DB] {
        mfrc.set_antenna_gain(gain).expect("set_antenna_gain");
        check_true!(mfrc.antenna_gain().unwrap_or(0xFF) == gain, "antenna_gain read back", passed, failed);
    }

    let _present = mfrc.is_card_present();

    mfrc.reset().expect("reset");
    check_true!(true, "reset accepted", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/rfid/mfrc522_test/main.go
//go:build linux && !tinygo

// MFRC522 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the MFRC522 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
//
// MFRC522 I2C address is 0x28 with all address pins tied LOW; set
// I2C_ADDR to override. The card-detection and read-block checks
// require a MIFARE Classic card placed in the field; the version,
// antenna, and self-test checks do not.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/rfid"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x28"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	tr, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr.Close()

	chip, err := rfid.NewMFRC522Full(tr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	// --- Identification and configuration ---
	chipType, ver, err := chip.Version()
	check("version_reads", err == nil)
	check("version_chip_type_0x09", chipType == 0x09)
	check("version_is_1_or_2", ver == 1 || ver == 2)

	gain, err := chip.AntennaGain()
	check("antenna_gain_default_0x40", err == nil && gain == rfid.RX_GAIN_33_DB)

	if err := chip.SetAntennaGain(rfid.RX_GAIN_48_DB); err != nil {
		fmt.Fprintln(os.Stderr, "set_antenna_gain:", err)
	}
	gain, _ = chip.AntennaGain()
	check("antenna_gain_48db", gain == rfid.RX_GAIN_48_DB)

	if err := chip.SetAntennaGain(rfid.RX_GAIN_33_DB); err != nil {
		fmt.Fprintln(os.Stderr, "set_antenna_gain:", err)
	}

	// --- Antenna toggling ---
	if err := chip.AntennaOff(); err != nil {
		fmt.Fprintln(os.Stderr, "antenna_off:", err)
	}
	if err := chip.AntennaOn(); err != nil {
		fmt.Fprintln(os.Stderr, "antenna_on:", err)
	}
	check("antenna_toggle_no_error", true)

	// --- Card detection (optional — only passes if a card is in field) ---
	present, err := chip.IsCardPresent()
	if err != nil {
		fmt.Fprintln(os.Stderr, "is_card_present:", err)
	}
	if present {
		uid, err := chip.ReadUID()
		check("read_uid_with_card", err == nil && uid != nil && (len(uid) == 4 || len(uid) == 7 || len(uid) == 10))
	} else {
		fmt.Println("note: no card in field; skipping read_uid check")
	}

	// --- Digital self test (no card needed) ---
	ok, err := chip.SelfTest()
	check("self_test", err == nil && ok)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
